### DISCUSSION TITLE
SSL Library Error: error:0A000126:SSL routines::unexpected eof while reading

### DISCUSSION THREAD

**Discussion by vinay-mummadi-digicert** [openssl/openssl#22690] [Q&A]
hi team, have been facing this error after upgrading to openssl 3.1.2. earlier  we were on openssl 1.1.1u and it was stable. this is the details from the log. 
```
[Tue Oct 17 05:58:16.863055 2023] [proxy:debug] [pid 26272:tid 140171114653440] proxy_util.c(2587): AH00943: AJP: has released connection for (m2be-q2-ap:9009)
[Tue Oct 17 05:58:16.996852 2023] [ssl:debug] [pid 26271:tid 140171156616960] ssl_engine_io.c(1151): [client IP4:55711] AH02001: Connection closed to child 15 with standard shutdown (server a.lab.net:443)
[Tue Oct 17 05:58:17.559642 2023] [ssl:info] [pid 26271:tid 140171165009664] [client IP6:5060] AH01964: Connection to child 14 established (server a.lab.net:443)
[Tue Oct 17 05:58:17.559786 2023] [ssl:info] [pid 26271:tid 140171165009664] (70014)End of file found: [client IP6:5060] AH02008: SSL library error 1 in handshake (server a.lab.net:443)
[Tue Oct 17 05:58:17.559806 2023] [ssl:info] [pid 26271:tid 140171165009664] SSL Library Error: error:0A000126:SSL routines::unexpected eof while reading
[Tue Oct 17 05:58:17.559811 2023] [ssl:info] [pid 26271:tid 140171165009664] [client IP6:5060] AH01998: Connection closed to child 14 with abortive shutdown (server a.lab.net:443)
[Tue Oct 17 05:58:18.424232 2023] [ssl:info] [pid 26271:tid 140171328984832] [client IP5:47602] AH01964: Connection to child 3 established (server a.lab.net:443)
[Tue Oct 17 05:58:18.424413 2023] [ssl:info] [pid 26271:tid 140171328984832] (70014)End of file found: [client IP5:47602] AH02008: SSL library error 1 in handshake (server a.lab.net:443)
[Tue Oct 17 05:58:18.424427 2023] [ssl:info] [pid 26271:tid 140171328984832] SSL Library Error: error:0A000126:SSL routines::unexpected eof while reading
```
as we're struck , i'm posting it in the QnA section
please suggest . please let me know if you need any additional information on this . thanks.

**Comment by mattcaswell** [comment]
This usually happens when the remote peer abruptly shuts down the connection without sending a "close_notify" alert which is what you are supposed to do.

1.1.1 and earlier treated this incorrectly. In such a case SSL_read() would fail and `SSL_get_error()` would return `SSL_ERROR_SYS` - but with nothing in `errno`. This was actually incorrect behaviour - it should be treated as `SSL_ERROR_SSL` with an appropriate error in the error stack. But making that change in 1.1.1 branch was too disruptive so it was introduced in 3.x instead.

So, this behaviour on behalf of the peer has probably always been there, but you are just noticing it now because how OpenSSL reacts to it has changed.

One option for dealing with this is to set the `SSL_OP_IGNORE_UNEXPECTED_EOF` option. See:

https://www.openssl.org/docs/man3.1/man3/SSL_set_options.html

This causes OpenSSL to treat an unexpected eof as if the peer had closed down gracefully - but make sure you read the docs above around truncation attacks.

If you have control over the remote peer, then another option is to change the remote peer so that it closes down connections gracefully.

**Comment by vinay-mummadi-digicert** [comment]
Thanks Matt for your response. Where is this setting required. is it in any of the config files? Please suggest. Thanks and regards, Vinay9008855944

**Comment by mattcaswell** [comment]
If this is in your own application, then you can modify it to call `SSL_set_options` on the SSL object.

Otherwise it is possible to set this via the OpenSSL config file. A minimal OpenSSL config file to turn this option on might look like this (untested):

````
config_diagnostics = 1

openssl_conf = default_conf

[ default_conf ]

ssl_conf = ssl_sect

[ssl_sect]

system_default = ssl_default_sect

[ssl_default_sect]
Options = SSL_OP_IGNORE_UNEXPECTED_EOF
````

**Comment by vinay-mummadi-digicert** [comment]
Thanks Matt. I changed the file openssl-3.1.2/apps/openssl.cnf . the other .cnf files were in demo and test dir’s and I ignored them. I added the options line as below. I’ve built the rpm and testing in progress.i think this is what you suggested. I’ll keep you posted .  Thanks again for your guidance.

**Comment by mattcaswell** [comment]
Better is just to modify the installed config file rather than change it in the source.

Or create a whole new OpenSSL config file specific for your application. Before starting your application ensure the `OPENSSL_CONF` environment variable is set and pointing to the location of the config file.

**Comment by vinay-mummadi-digicert** [comment]
Hi Matt,   we tried the changes in the installed config file, restarted the httpd service which uses this openssl and getting the same errors. We added these as suggested.

````
openssl_conf = default_conf
[ default_conf ]
ssl_conf = ssl_sect
[ssl_sect] system_default = ssl_default_sect
[ssl_default_sect]
Options = SSL_OP_IGNORE_UNEXPECTED_EOF
````

The .cnf file also has this entry # Use this in order to automatically load providers.openssl_conf = openssl_init we tried with both  openssl_init, default_conf values for openssl_conf.  is there any way we can echo the value used for options to confirm that out settings are taken into account? please suggest  thanks,Vinay9008855944

**Comment by mattcaswell** [comment]
If you are still seeing the errors then the config option did not take effect. This could be for one of three reasons that I can think of:

1) There is an error in the config file that prevents the changes from taking effect

2) The application is using some other config file than the one that you changed

3) The application itself is overriding the config options or explicitly causing the config file not to be loaded.

For (1) you can zip the config file and attach it here if you like. I can take a look to see if it seems sane.

For (2) I am assuming you are using the "system" OpenSSL installation and you are modifying the system OpenSSL config file. You should check whether the application is using the system OpenSSL or its own one. You might also want to check whether the application has some custom `OPENSSL_CONF` environment variable setting pointing to some other config file.

For (3) you would have to raise an issue with whoever supplied your application.

**Comment by vinay-mummadi-digicert** [comment]
I’ve attached the cnf file. The changes are from line 39.Pls note that line 17 I’ve commented the “openssl_conf = “ line as it’s repeated after line 39 This is overview of our current appln. I generate the openssl rpm, with changes related to our app path. I then generate the apache httpd related rpm pointing to this new openssl. When the httpd service is started for our webservices framework, the errors are noted. Pls let me know if you need any more details. Thanks-- Vinay9008855944

**Comment by vinay-mummadi-digicert** [comment]
Hi Matt, Did you get a chance to review the .cnf file. pls suggest corrections if possible. We’ve noted this additional error in the log.Additional InformationErrors : index="ent" "SSL library error 1 in handshake" "ws. auth.com:443"Customre getting following errorsTC_HttpCommunication::process()[TC_HttpCommunication.cpp:934]:## Error : The connection with the server has been reset or terminated, or an incompatible SSL protocol was encountered. For example, WinHTTP version 5.1 does not support SSL2 unless the client specifically enables it Thanks, Vinay9008855944

**Comment by t8m** [comment]
@vinay-mummadi-digicert unfortunately your two last comments which you posted by e-mail are practically illegible due to formatting issues. Please do not use e-mail or possibly this particular e-mail client as it completely messes up the formatting and it is completely unclear what is the new text you've posted and what is the original comment you're replying to.

**Comment by vinay-mummadi-digicert** [comment]
understood. i'll start posting in the ticket.  i've added both 1,2 mails here.
[cnf.zip](https://github.com/openssl/openssl/files/13248640/cnf.zip)
[mattcaswell](https://github.com/mattcaswell)

part 1: I’ve attached the cnf file. 
The changes are from line 39.Pls note that line 17 I’ve commented the “openssl_conf = “ line as it’s repeated after line 39 
This is overview of our current appln. 
I generate the openssl rpm, with changes related to our app path. I then generate the apache httpd related rpm pointing to this new openssl. When the httpd service is started for our webservices framework, the errors are noted. Pls let me know if you need any more details. 

part 2: We’ve noted this additional error in the log.Additional InformationErrors : index="ent" "SSL library error 1 in handshake" "ws. auth.com:443"Customre getting following errorsTC_HttpCommunication::process()[TC_HttpCommunication.cpp:934]:## Error : The connection with the server has been reset or terminated, or an incompatible SSL protocol was encountered. For example, WinHTTP version 5.1 does not support SSL2 unless the client specifically enables it 

Thanks-- Vinay
9008855944

**Comment by mattcaswell** [comment]
> The changes are from line 39.Pls note that line 17 I’ve commented the “openssl_conf = “ line as it’s repeated after line 39
This is overview of our current appln.

The config file is not quite right. You've pasted the "openssl_conf =" line into the "[ new_oids ]" section - so it won't get noticed. You need to move all the stuff you pasted in from line 39 to be at line 32 instead.

**Comment by vinay-mummadi-digicert** [comment]
i moved the content to above "[ new_oids ]" . we're observing the same error. 
any option to debug or print the settings, to see if the .cnf file changes are incorporated ?

any other changes to try, please suggest.

**Comment by AntonZhernosek** [comment]
Having a similar issue connecting to a sql server from a dockerized container with Debian 12 and OpenSSL 3.0.11
Would appreciate the feedback very much because the errors I get look related. I've tried the suggested changes to the .cnf above and it doesn't seem to change anything

**Comment by gbryant200** [comment]
I'm having the same issue on Ubuntu 22.04

### openssl version -a
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
built on: Wed Jan 31 18:43:23 2024 UTC
platform: debian-amd64
options:  bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -Wa,--noexecstack -g -O2 -ffile-prefix-map=/build/openssl-l59e1T/openssl-3.0.2=. -flto=auto -ffat-lto-objects -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -DOPENSSL_TLS_SECURITY_LEVEL=2 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG -Wdate-time -D_FORTIFY_SOURCE=2
OPENSSLDIR: "/usr/lib/ssl"
ENGINESDIR: "/usr/lib/x86_64-linux-gnu/engines-3"
MODULESDIR: "/usr/lib/x86_64-linux-gnu/ossl-modules"
Seeding source: os-specific
CPUINFO: OPENSSL_ia32cap=0xfeda32035f8bffff:0x800d19e67a9

It seems like the `Options = SSL_OP_IGNORE_UNEXPECTED_EOF` isn't doing anything.

I tried to run this from the command line: `SSL_CONF_cmd(ctx,"Options", "SSL_OP_IGNORE_UNEXPECTED_EOF")` but received and unexpected token ctx, "Options"

this fails as well:
`openssl s_client -ignore_unexpected_eof -connect github.com:443`

from what I can tell, at least for me, this only happens connecting to github.

Any ideas/help would be GREATLY appreciated.

**Reply by Florida-Rob** [reply]
has anyone found a fix for EOF with OpenSSL 3.x?
seeing the same with Mosquitto, in Ubuntu 22.04

**Comment by jrella1** [comment]
I too am having this issue, and I have no experience modifying the openssl config file. I've attached the file, if someone could assist?
[openssl.cnf.txt](https://github.com/openssl/openssl/files/15485187/openssl.cnf.txt)

**Comment by singla2011** [comment]
We are also facing same issue. Is there any fix available for this EOF with OpenSSL 3.x or any Workaround available?

**Comment by mattcaswell** [comment]
The solution is mentioned earlier in this thread:

https://github.com/openssl/openssl/discussions/22690#discussioncomment-7534911

**Reply by singla2011** [reply]
We tried this but it did not work.

**Reply by singla2011** [reply]
We have tried this solution but it did not work. Can you provide any other resolution for this issue.

**Comment by KisCserkesz** [comment]
I have the same problem with mosquitto_pub, OpenSSl 3.0.2, on Kubuntu 22.04.
The config file modification did nothing. I think msoquitto_pub has an internal setting... but I have to check the source...

**Comment by mteall** [comment]
@mattcaswell you seem very familiar with this issue, I figure it's worth a hail Mary in asking you about this same situation:

When upgrading a K8s based php service from PHP 8.0 to PHP 8.2 (which upgraded openssl from 1.1.1 to 3.3.2, and curl from 8.4.0 to 8.9.1) I started getting the following error: "curl: (56) OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 0" logged from my app.

I then figured out how to reproduce this issue on a local container using both curl and openssl client via command line.  Curl via command line resulted in a 200 OK response, with the expected response body, and the error: curl: (56) OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 0.

Using openssl s_client -connect, then pasiting my request:
openssl s_client -connect the.host.site:443 -quiet
....
POST /site/url HTTP/1.1
Content-Length: 517
Host: the.host.site
User-Agent: MyServiceUserAgentString
Authorization: Basic ASDF1234=
Content-Type: application/json

{"key1":"val1"}

I received the following error:
"283BC777B57F0000:error:0A000126:SSL routines::unexpected eof while reading:ssl/record/rec_layer_s3.c:687:"

I then found your post and implemented your suggestion from this page: https://github.com/openssl/openssl/discussions/24810

Then re-running my post with openssl s_client connect from command line the error message was gone.  However with curl, the curl error 56, openssl errorno 0 issue still remained.  As an experiment, I messed up the openssl.cnf config by changing:

openssl_conf = default_conf
to
openssl_conf = default_conf123

As expected re-running openssl s_client connect failed immediately because of the config issue.  However when I re-ran my curl command, I received a response, with the same darn curl error 56, openssl errno 0!  This is leading me to believe that when curl is using openssl as the backend, it is NOT reading from the openssl config file!  I further confirmed this by removing openssl entirely via apk del openssl, then re-running curl and getting the same error. I'm starting to think THIS is why a some people were commenting that the suggested config change wasn't working for them (on various threads).

My php service is built on top of the following docker image, if this helps: php:8.2-fpm-alpine3.20

**Reply by mattcaswell** [reply]
> This is leading me to believe that when curl is using openssl as the backend, it is NOT reading from the openssl config file! 

This is entirely possible. In an earlier post in this thread I noted some reasons why the config settings might not stick:

https://github.com/openssl/openssl/discussions/22690#discussioncomment-7534915

Whether or not curl overrides things I do not know.

**Reply by KisCserkesz** [reply]
Agree...
The real problem is at client software side, cURL in this case.
I have tried using different MQTT tools (Hive MQTT, MQTTX) instead of cURL, with the same underlaying openSSL libraries.
The result: they have worked. All of them. But cURL does not. Surely it does not handle this new (and corrected) behaviour introduced in openSSL3.
With cURL I have not found any workaround.

**Comment by AposLaz** [comment]
I am facing the same error when applying `git clone` from a BitBucket repository on Unix-based systems like Ubuntu, Red Hat, etc. However, when I use my Windows machine, I do not encounter the same issue, and git clone completes successfully. I understand that Git uses Schannel on Windows instead of OpenSSL for SSL communication, but I am not sure what the differences between them.

I followed the advice from this page: https://github.com/openssl/openssl/discussions/24810 

I attached the `openssl.conf` file to confirm that it follows the suggested configuration:
[openssl.cnf.txt](https://github.com/user-attachments/files/17199854/openssl.cnf.txt)

OpenSSL version: `3.0.7`, 1 Nov 2022

Additionally, I tried using a previous version of OpenSSL, but had no success. OpenSSL version: `1.1.1`, 11 Sep 2018

However the issue remains unresolved. Any suggestions about it?

**Comment by mteall** [comment]
This probably won't help anyone, but in my case, downgrading my docker image from php:8.2-fpm-alpine3.20 to php:8.2-fpm-alpine3.19 got rid of my immediate problem (curl err 56, openssl errno 0).  They both shipped with curl 8.9.1 (alpine3.20 just bumped that to 8.10.1), but the alpine3.20 image comes with openssl 3.3.2 and alpine3.19 comes with openssl 3.1.7.

The interesting part is that I still see the enexpected EOF error when making the call via openssl s_client -connect, but curl doesn't seem to notice or report any issue.

Another thing I learned is that the configuration when alpine compiles openssl and curl isn't exactly vanilla... I have been burned by alpine more times than I can count now.  Not sure it's worth all the hassle just to have "small images".

**Comment by ensingerphilipp** [comment]
Also facing this issues with Openssl 3.0.13 when using a dockerized Container

**Comment by jdsampayo** [comment]
I encountered the `SSL routines::unexpected eof` error because the domain I was connecting to randomly resolved to two different IP addresses. I’m not entirely sure why, perhaps it was a caching issue, but one of these IPs was out of service. As a result, whenever the domain resolved to that problematic IP, the connection failed with an EOF error.

As a temporary workaround, I modified `/etc/hosts` to force the domain to always resolve to the working IP address.

If you are experiencing similar intermittent “EOF” or “connection reset” errors, check whether your domain is resolving to multiple IP addresses, this could be on some redirects, on my case it used subdomains like ww22, ww28, ww30, etc, some of which may be sending to the offline IP or not properly configured. If that’s the case, you can temporarily override the IP in `/etc/hosts` for all subdomains. However, the more permanent solution would be to fix your DNS settings or adjust your server configuration so that all IP addresses are valid.

I initially tried using the `openssl.conf` solution, but adding it caused another error (code 6), before I knew about the offline IP. Therefore, I removed that configuration file. However, I did notice that a line in the shared config had changed:

```
Options = SSL_OP_IGNORE_UNEXPECTED_EOF
```

should now be

```
Options = IgnoreUnexpectedEOF
```

It might have changed after @mattcaswell shared his config file. Here’s an example of the updated configuration:

```
config_diagnostics = 1
openssl_conf = default_conf

[ default_conf ]
ssl_conf = ssl_sect

[ssl_sect]
system_default = ssl_default_sect

[ssl_default_sect]
Options = IgnoreUnexpectedEOF
```

Remember to set the `OPENSSL_CONF` environment variable to point to your configuration file if you need this workaround.

**Reply by tsogbadrakh-qp** [reply]
I have tried to create a config file and set `OPENSSL_CONF` env in my docker image as follows. But it doesn't resolve the issue.
Maybe it's because https://github.com/ruby/openssl. I didn't dig their implementation regarding the option.

```dockerfile
# debian bookworm
COPY ./dockerfiles/openssl.conf /etc/ssl/openssl.cnf
ENV OPENSSL_CONF=/etc/ssl/openssl.cnf
```

**Comment by valera-rozuvan** [comment]
So the problem seems to have gone away with recent versions of `openssl`.

Consider the following script:

```shell
#!/bin/bash

HostDomain="google.com" ; \
HostIP="216.58.215.110" ; \
HostPort="443" ; \
dataToPost='{ data: "" }' ; \
dataLen=$(expr length "${dataToPost}" ) ; \
  ( printf "POST /test HTTP/1.1\r\n" ; \
    printf "Host: %s\r\n" "${HostDomain}" ; \
    printf "User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:58.0) Gecko/20100101 Firefox/58.0\r\n" ; \
    printf "Connection: close\r\n" ; \
    printf "Content-Length: %d\r\n" "$((dataLen+1))" ; \
    printf "Content-Type: application/x-www-form-urlencoded\r\n" ; \
    printf "\r\n" ; \
    printf "%s" "${dataToPost}" ; \
    sleep 1.5 ) | openssl s_client -connect ${HostIP}:${HostPort}
```

When running this - I get the expected response, but the last line is the infamous error:

```text
40F7C6A1FD7E0000:error:0A000126:SSL routines::unexpected eof while reading:../ssl/record/rec_layer_s3.c:696:
```

Now if I add the flag `-ignore_unexpected_eof`:

```shell
#!/bin/bash

HostDomain="google.com" ; \
HostIP="216.58.215.110" ; \
HostPort="443" ; \
dataToPost='{ data: "" }' ; \
dataLen=$(expr length "${dataToPost}" ) ; \
  ( printf "POST /test HTTP/1.1\r\n" ; \
    printf "Host: %s\r\n" "${HostDomain}" ; \
    printf "User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:58.0) Gecko/20100101 Firefox/58.0\r\n" ; \
    printf "Connection: close\r\n" ; \
    printf "Content-Length: %d\r\n" "$((dataLen+1))" ; \
    printf "Content-Type: application/x-www-form-urlencoded\r\n" ; \
    printf "\r\n" ; \
    printf "%s" "${dataToPost}" ; \
    sleep 1.5 ) | openssl s_client -ignore_unexpected_eof -connect ${HostIP}:${HostPort}
```

The error is no more!

---

My system:

```text
$ cat /etc/*-release
PRETTY_NAME="Debian GNU/Linux 13 (trixie)"
NAME="Debian GNU/Linux"
VERSION_ID="13"
VERSION="13 (trixie)"
VERSION_CODENAME=trixie
DEBIAN_VERSION_FULL=13.1
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

$ uname -a
Linux lp-rusty 6.12.43+deb13-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.12.43-1 (2025-08-27) x86_64 GNU/Linux

$ openssl --version
OpenSSL 3.5.1 1 Jul 2025 (Library: OpenSSL 3.5.1 1 Jul 2025)
```
